pytrip/concave_tool.py

```diff
@@ -0,0 +1,302 @@
+                part_b_to_merge = part_b
+
+    return a_reverse, b_reverse, part_a_to_merge, part_b_to_merge
+
+
+def connect_two_parts(a_reverse, b_reverse, part_a_to_merge, part_b_to_merge):
+    """
+    Connects two parts of contour depending on which ends of parts should be used.
+    """
+    #   True - connect starting from the end of array
+    #   False - connect starting from the start of array
+    if not a_reverse and not b_reverse:
+        # adding two parts with their starts, must reverse part_a and then append part_b
+        part_a_to_merge.data = list(reversed(part_a_to_merge.data))
+        part_a_to_merge.data.extend(part_b_to_merge.data)
+    elif not a_reverse and b_reverse:
+        # adding to a start b end, so we can add a to b, and then assign b to a
+        part_b_to_merge.data.extend(part_a_to_merge.data)
+        part_a_to_merge.data = part_b_to_merge.data
+    elif a_reverse and not b_reverse:
+        # adding to a end b start, we just need to extend a
+        part_a_to_merge.data.extend(part_b_to_merge.data)
+    elif a_reverse and b_reverse:
+        # adding to a end b end, we need to reverse b
+        part_a_to_merge.data.extend(reversed(part_b_to_merge.data))
+
+
+def group_points_for_multiple_contours(points_lists):
+    """
+    Divides points into groups that will be used to construct contour for each one.
+    """
+    # empty array for each group/contour
+    multiple_contours = []
+    # empty array for current group/contour
+    current_contour = []
+
+    for points_list in points_lists:
+        # if current list of points (intersection) is not empty, append that list to current group
+        if len(points_list) > 0:
+            current_contour.append(points_list)
+        # if it is empty it means that current contour has ended
+        else:
+            # save current contour as one of multiple ones
+            multiple_contours.append(current_contour)
+            # create new empty group/contour
+            current_contour = []
+    # save the last contour if it exists
+    if current_contour:
+        multiple_contours.append(current_contour)
+    # remove empty ones
+    #   it could be done earlier by filtering list passed as parameter
+    multiple_contours = [contour for contour in multiple_contours if contour]
+    # return only non-empty groups/contours
+    return multiple_contours
+
+
+# -------------------- end block of logic methods --------------------
+
+
+# -------------------- start block of wrapper/helper methods --------------------
+def create_contour_parts(points_lists):
+    mapped_points = list(map(map_points_to_special_points, points_lists))
+
+    parts_of_contour = []
+
+    for current_points in mapped_points:
+        append_points_to_parts(current_points, parts_of_contour)
+        create_parts_for_loose_points(current_points, parts_of_contour)
+
+    return parts_of_contour
+
+
+def connect_all_parts(parts):
+    while len(parts) > 1:

```

Comment:
Are we sure that this `while` doesn't enter endless loop ?
I'd suggest some guardian, like a counter set to very large number like 10^6, if the counter is exceeded, then we end a loop and issue warning

Reply:
Yes we are, because we search for the closest pair od parts and merge them, so the list is getting shorter and shorter.